### DISCUSSION TITLE
Low Gravity + Restitution - Funny Floating/Freezing

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#106] [Q&A] [answered]
I was debugging this for hours in my own library, but now realize it also happens in the stock one.

![2024-02-17-jolt-floating-objects](https://github.com/jrouwe/JoltPhysics.js/assets/756488/ae4ad06e-03d5-42ed-b0e1-b9f2d3f015fa)

Do we know if this is predictable, or a hook that can be used to force gravity to re-take objects once they start to float?
This seems to happen more frequently in my own tests, but below reproduces it. Sometimes 1x ball, sometimes all of them, sometimes just the 2x big ones.
```javascript
    import initJolt from './js/jolt-physics.wasm-compat.js';

    initJolt().then(function (Jolt) {
      // Initialize this example
      initExample(Jolt, null);

      const data = [
        { "position": [0.799, 5.37, -0.460], "radius": 0.5, "shape": "Sphere" },
        { "position": [-0.410, 5.37, 0.679], "radius": 1.0, "shape": "Sphere" },
        { "position": [1.449, 5.37, -0.460], "radius": 0.25, "shape": "Sphere" },
        { "position": [0.239, 5.37, 0.679], "radius": 0.25, "shape": "Sphere" },
        { "position": [0.239, 5.37, -0.460], "radius": 0.25, "shape": "Sphere" },
        { "position": [0.799, 5.37, 0.140], "radius": 0.25, "shape": "Sphere" },
        { "position": [-0.410, 5.37, 0.140], "radius": 0.25, "shape": "Sphere" },
        { "position": [0.239, 5.37, 0.140], "radius": 0.25, "shape": "Sphere" },
        { "position": [0, 0, 0], "shape": "Box" }];

      const position = new Jolt.Vec3();
      const worldRotation = new Jolt.Quat(0, 0, 0, 1);
      const extent = new Jolt.Vec3();
      data.forEach(item => {
        position.Set(...item.position);
        switch (item.shape) {
          case 'Box':
            extent.Set(7.5, 0.25, 7.5);
            const box = createBox(position, worldRotation, extent, Jolt.EMotionType_Static, LAYER_NON_MOVING, 0xCCCCCC)
            box.SetRestitution(0.5);
            break;
          case 'Sphere':
            const sphere = createSphere(position, item.radius/2, Jolt.EMotionType_Dynamic, LAYER_MOVING, 0xFF0000);
            break;
        }
        physicsSystem.SetGravity(new Jolt.Vec3(0, -.5, 0));
      });
    });
```

I noticed it when I was trying to confirm syncing models in a nested hierarchy with layered transforms, so originally thought it was that, until I found out that minimal use-case above also triggered it.
One the perk side, I'm getting a bit better at making my own debug views... I think I will skip over the fact I can't handle skew on models for the center-most scaled-group Prefab box.
![2024-02-17-jolt-floating-debug](https://github.com/jrouwe/JoltPhysics.js/assets/756488/28045dab-0005-4a39-ac8d-8295db738b64)

**Comment by jrouwe** [comment, marked as answer]
I think what you're seeing here is that some of the spheres move so slow for a prolonged period of time that they fall asleep to conserve CPU time. ```PhysicsSettings::mTimeBeforeSleep``` (default: 0.5) has been tweaked for earth gravity. In this low gravity condition you probably want to make it higher.

**Reply by PhoenixIllusion** [reply]
That was it. It seems fine to change that on the fly, but currently due to the [Value] nature it needs to be done:
```javascript
        physicsSystem.SetGravity(new Jolt.Vec3(0, -.5, 0));
        const settings = physicsSystem.GetPhysicsSettings();
        settings.mTimeBeforeSleep = 1.5;
        physicsSystem.SetPhysicsSettings(settings);
```

Since it is a function, we can probably make that a `[Const, Ref]`
